File: backend/app/routes/clubs.py
import uuid
from datetime import date

from flask import Blueprint, jsonify, request
from flask_jwt_extended import jwt_required

from app.decorators import role_required
from app.extensions import db
from app.models import Club, User, ClubMember
from app.serializers import club_to_dict, user_to_dict

bp = Blueprint("clubs", __name__, url_prefix="/api/clubs")


def _resolve_head_id(raw: str | None) -> str:
    if not raw or raw == "admin":
        admin = User.query.filter_by(role="admin").first()
        return admin.id if admin else ""
    return raw


@bp.route("", methods=["GET"])
def list_clubs():
    clubs = Club.query.order_by(Club.name).all()
    return jsonify([club_to_dict(c) for c in clubs]), 200


@bp.route("/my", methods=["GET"])
@jwt_required()
def list_my_clubs():
    from flask_jwt_extended import get_jwt_identity
    uid = get_jwt_identity()
    memberships = ClubMember.query.filter_by(user_id=uid).all()
    club_ids = [m.club_id for m in memberships]
    clubs = Club.query.filter(Club.id.in_(club_ids)).all() if club_ids else []
    return jsonify([club_to_dict(c) for c in clubs]), 200


@bp.route("/<cid>/join", methods=["POST"])
@jwt_required()
def join_club(cid):
    from flask_jwt_extended import get_jwt_identity
    uid = get_jwt_identity()
    user = db.session.get(User, uid)
    if not user:
        return jsonify({"message": "Unauthorized"}), 401
    
    club = db.session.get(Club, cid)
    if not club:
        return jsonify({"message": "Club not found"}), 404
    
    existing = ClubMember.query.filter_by(user_id=uid, club_id=cid).first()
    if existing:
        return jsonify({"message": "Already a member"}), 400
    
    db.session.add(ClubMember(user_id=uid, club_id=cid))
    club.member_count = (club.member_count or 0) + 1
    db.session.commit()
    
    return jsonify({"message": "Joined successfully"}), 200


@bp.route("/<cid>", methods=["GET"])
def get_club(cid):
    c = db.session.get(Club, cid)
    if not c:
        return jsonify({"message": "Not found"}), 404
    return jsonify(club_to_dict(c)), 200


@bp.route("/<cid>/members", methods=["GET"])
@jwt_required()
def list_club_members(cid):
    from flask_jwt_extended import get_jwt_identity
    uid = get_jwt_identity()
    user = db.session.get(User, uid)
    if not user:
        return jsonify({"message": "Unauthorized"}), 401
    
    club = db.session.get(Club, cid)
    if not club:
        return jsonify({"message": "Not found"}), 404
    
    # Permission check:
    # Admin can see all
    # Club Head can see their own
    # Student can see if they are a member
    is_admin = user.role == "admin"
    is_head = user.role == "club_head" and user.club_id == cid
    membership = ClubMember.query.filter_by(user_id=uid, club_id=cid).first()
    is_member = membership is not None
    
    if not (is_admin or is_head or is_member):
        return jsonify({"message": "Forbidden"}), 403
    
    memberships = ClubMember.query.filter_by(club_id=cid).all()
    user_ids = [m.user_id for m in memberships]
    users = User.query.filter(User.id.in_(user_ids)).all() if user_ids else []
    
    return jsonify([user_to_dict(u) for u in users]), 200


@bp.route("", methods=["POST"])
@jwt_required()
@role_required("admin")
def create_club():
    data = request.get_json(silent=True) or {}
    name = (data.get("name") or "").strip()
    if not name:
        return jsonify({"message": "Name required"}), 400

    head_id = _resolve_head_id(data.get("headId") or data.get("head_id"))
    if not head_id:
        return jsonify({"message": "Could not resolve club head"}), 400

    cid = f"club-{uuid.uuid4().hex[:12]}"
    today = date.today().isoformat()
    club = Club(
        id=cid,
        name=name,
        description=data.get("description") or "",
        category=data.get("category") or "",
        member_count=0,
        head_id=head_id,
        created_at=today,
        logo=data.get("logo"),
    )
    db.session.add(club)
    
    head_user = db.session.get(User, head_id)
    if head_user:
        head_user.club_id = cid
        head_user.role = "club_head"

    db.session.commit()
    return jsonify(club_to_dict(club)), 201


@bp.route("/<cid>", methods=["PUT"])
@jwt_required()
@role_required("admin", "club_head")
def update_club(cid):
    from flask_jwt_extended import get_jwt_identity
    uid = get_jwt_identity()
    user = db.session.get(User, uid)
    if not user:
        return jsonify({"message": "Unauthorized"}), 401

    c = db.session.get(Club, cid)
    if not c:
        return jsonify({"message": "Not found"}), 404

    if user.role != "admin" and (user.role != "club_head" or user.club_id != cid):
        return jsonify({"message": "Forbidden"}), 403

    data = request.get_json(silent=True) or {}

    if user.role != "admin":
        # Club heads can only update description and logo
        if "description" in data:
            c.description = data["description"] or ""
        if "logo" in data:
            c.logo = data.get("logo")
    else:
        # Admin can update name, description, category, headId, logo
        if "name" in data:
            c.name = data["name"]
        if "description" in data:
            c.description = data["description"] or ""
        if "category" in data:
            c.category = data["category"] or ""
        if "headId" in data or "head_id" in data:
            hid = data.get("headId") or data.get("head_id")
            new_head_id = _resolve_head_id(hid) or c.head_id
            if new_head_id != c.head_id:
                old_head = db.session.get(User, c.head_id)
                if old_head and old_head.club_id == c.id:
                    old_head.club_id = None
                c.head_id = new_head_id
                new_head = db.session.get(User, new_head_id)
                if new_head:
                    new_head.club_id = c.id
                    new_head.role = "club_head"
                
                if old_head:
                    other_heads = Club.query.filter(Club.head_id == old_head.id, Club.id != cid).first()
                    if not other_heads and old_head.role == "club_head":
                        old_head.role = "student"

        if "logo" in data:
            c.logo = data.get("logo")

    db.session.commit()
    return jsonify(club_to_dict(c)), 200


@bp.route("/<cid>", methods=["DELETE"])
@jwt_required()
@role_required("admin")
def delete_club(cid):
    from app.models import ClubMember, Event, User

    c = db.session.get(Club, cid)
    if not c:
        return jsonify({"message": "Not found"}), 404

    ClubMember.query.filter_by(club_id=cid).delete(synchronize_session=False)
    User.query.filter_by(club_id=cid).update({"club_id": None}, synchronize_session=False)
    Event.query.filter_by(club_id=cid).delete(synchronize_session=False)

    db.session.delete(c)
    db.session.commit()
    return "", 204

File: backend/src/routes/clubs.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');

const { jwtRequired, roleRequired } = require('../middleware/auth');
const User = require('../models/User');
const Club = require('../models/Club');
const ClubMember = require('../models/ClubMember');
const Event = require('../models/Event');
const { clubToDict, userToDict } = require('../serializers');

const router = express.Router();

async function resolveHeadId(raw) {
  if (!raw || raw === 'admin') {
    const admin = await User.findOne({ role: 'admin' });
    return admin ? admin._id : '';
  }
  return raw;
}

router.get('/', async (_req, res) => {
  const clubs = await Club.find().sort({ name: 1 });
  res.status(200).json(await Promise.all(clubs.map(clubToDict)));
});

router.get('/my', jwtRequired, async (req, res) => {
  const memberships = await ClubMember.find({ userId: req.userId });
  const clubIds = memberships.map((m) => m.clubId);
  const clubs = clubIds.length ? await Club.find({ _id: { $in: clubIds } }) : [];
  res.status(200).json(await Promise.all(clubs.map(clubToDict)));
});

router.post('/:cid/join', jwtRequired, async (req, res) => {
  const { cid } = req.params;
  const user = await User.findById(req.userId);
  if (!user) {
    return res.status(401).json({ message: 'Unauthorized' });
  }

  const club = await Club.findById(cid);
  if (!club) {
    return res.status(404).json({ message: 'Club not found' });
  }

  const existing = await ClubMember.findOne({ userId: req.userId, clubId: cid });
  if (existing) {
    return res.status(400).json({ message: 'Already a member' });
  }

  await new ClubMember({ userId: req.userId, clubId: cid }).save();
  club.memberCount = (club.memberCount || 0) + 1;
  await club.save();

  res.status(200).json({ message: 'Joined successfully' });
});

router.get('/:cid', async (req, res) => {
  const c = await Club.findById(req.params.cid);
  if (!c) {
    return res.status(404).json({ message: 'Not found' });
  }
  res.status(200).json(await clubToDict(c));
});

router.get('/:cid/members', jwtRequired, async (req, res) => {
  const { cid } = req.params;
  const user = await User.findById(req.userId);
  if (!user) {
    return res.status(401).json({ message: 'Unauthorized' });
  }

  const club = await Club.findById(cid);
  if (!club) {
    return res.status(404).json({ message: 'Not found' });
  }

  const isAdmin = user.role === 'admin';
  const isHead = user.role === 'club_head' && user.clubId === cid;
  const membership = await ClubMember.findOne({ userId: req.userId, clubId: cid });
  const isMember = !!membership;

  if (!(isAdmin || isHead || isMember)) {
    return res.status(403).json({ message: 'Forbidden' });
  }

  const memberships = await ClubMember.find({ clubId: cid });
  const userIds = memberships.map((m) => m.userId);
  const users = userIds.length ? await User.find({ _id: { $in: userIds } }) : [];

  res.status(200).json(await Promise.all(users.map(userToDict)));
});

router.post('/', jwtRequired, roleRequired('admin'), async (req, res) => {
  const data = req.body || {};
  const name = (data.name || '').trim();
  if (!name) {
    return res.status(400).json({ message: 'Name required' });
  }

  const headId = await resolveHeadId(data.headId || data.head_id);
  if (!headId) {
    return res.status(400).json({ message: 'Could not resolve club head' });
  }

  const cid = `club-${uuidv4().replace(/-/g, '').slice(0, 12)}`;
  const today = new Date().toISOString().slice(0, 10);
  const club = new Club({
    _id: cid,
    name,
    description: data.description || '',
    category: data.category || '',
    memberCount: 0,
    headId,
    createdAt: today,
    logo: data.logo || null,
  });
  await club.save();

  const headUser = await User.findById(headId);
  if (headUser) {
    headUser.clubId = cid;
    headUser.role = 'club_head';
    await headUser.save();
  }

  res.status(201).json(await clubToDict(club));
});

router.put('/:cid', jwtRequired, roleRequired('admin', 'club_head'), async (req, res) => {
  const { cid } = req.params;
  const user = await User.findById(req.userId);
  if (!user) {
    return res.status(401).json({ message: 'Unauthorized' });
  }

  const c = await Club.findById(cid);
  if (!c) {
    return res.status(404).json({ message: 'Not found' });
  }

  if (user.role !== 'admin' && (user.role !== 'club_head' || user.clubId !== cid)) {
    return res.status(403).json({ message: 'Forbidden' });
  }

  const data = req.body || {};

  if (user.role !== 'admin') {
    if ('description' in data) c.description = data.description || '';
    if ('logo' in data) c.logo = data.logo || null;
  } else {
    if ('name' in data) c.name = data.name;
    if ('description' in data) c.description = data.description || '';
    if ('category' in data) c.category = data.category || '';
    if ('headId' in data || 'head_id' in data) {
      const hid = data.headId || data.head_id;
      const newHeadId = (await resolveHeadId(hid)) || c.headId;
      if (newHeadId !== c.headId) {
        const oldHead = await User.findById(c.headId);
        if (oldHead && oldHead.clubId === c._id) {
          oldHead.clubId = null;
        }
        c.headId = newHeadId;
        const newHead = await User.findById(newHeadId);
        if (newHead) {
          newHead.clubId = c._id;
          newHead.role = 'club_head';
          await newHead.save();
        }
        if (oldHead) {
          const otherHeads = await Club.findOne({ headId: oldHead._id, _id: { $ne: cid } });
          if (!otherHeads && oldHead.role === 'club_head') {
            oldHead.role = 'student';
          }
          await oldHead.save();
        }
      }
    }
    if ('logo' in data) c.logo = data.logo || null;
  }

  await c.save();
  res.status(200).json(await clubToDict(c));
});

router.delete('/:cid', jwtRequired, roleRequired('admin'), async (req, res) => {
  const { cid } = req.params;
  const c = await Club.findById(cid);
  if (!c) {
    return res.status(404).json({ message: 'Not found' });
  }

  await ClubMember.deleteMany({ clubId: cid });
  await User.updateMany({ clubId: cid }, { $set: { clubId: null } });
  await Event.deleteMany({ clubId: cid });
  await c.deleteOne();

  res.status(204).send();
});

module.exports = router;
